diff --git a/pipeline/preprocessing.py b/pipeline/preprocessing.py
index 932ae0b..d8ab14d 100644
--- a/pipeline/preprocessing.py
+++ b/pipeline/preprocessing.py
@@ -138,8 +138,6 @@ if stage == 2:
 
 # Parameters
 parameters = dict()
-model_arch = "TABL"
-parameters["arch"] = model_arch
 parameters["name"] = model
 parameters["WL"] = 200  # WINDOW LONG
 parameters["pt"] = 1
diff --git a/temp/cp_end.ckpt b/temp/cp_end.ckpt
index c1c1d9e..e682868 100644
Binary files a/temp/cp_end.ckpt and b/temp/cp_end.ckpt differ
diff --git a/temp/is_script_finished.txt b/temp/is_script_finished.txt
index 40ca0d3..ae2eeda 100644
--- a/temp/is_script_finished.txt
+++ b/temp/is_script_finished.txt
@@ -1 +1 @@
-full_script_time64.39192986488342
\ No newline at end of file
+start_script_time1589149219.5685346
\ No newline at end of file
Submodule third_party_libraries/TABL 7f25314..df3e92d:
diff --git a/third_party_libraries/TABL/__pycache__/Layers.cpython-37.pyc b/third_party_libraries/TABL/__pycache__/Layers.cpython-37.pyc
new file mode 100644
index 0000000..12f31e7
Binary files /dev/null and b/third_party_libraries/TABL/__pycache__/Layers.cpython-37.pyc differ
diff --git a/third_party_libraries/TABL/__pycache__/Models.cpython-37.pyc b/third_party_libraries/TABL/__pycache__/Models.cpython-37.pyc
new file mode 100644
index 0000000..d8950e3
Binary files /dev/null and b/third_party_libraries/TABL/__pycache__/Models.cpython-37.pyc differ
